Pm regions (#1962)
NordicPlayground/fw-nrfconnect-nrf: scripts/partition_manager/partition_manager.rst

   #endif /* CONFIG_SOC_NRF9160 */

.. _pm_regions:

Regions
=======

The Partition Manager places partitions in different *regions*.
For example, you can use regions for internal flash memory and external flash memory.

> These regions are predefined as ``flash_primary`` and ``external_flash``, respectively.

Reviewer: Is that correct?

Author: Yes, you can use these references in your `pm.yml`. The external flash will only work if you have set the corresponding kconfig option.

Author: But as all of this is explained better in the "EXternal flash" chapter, IMO this does not give any value, so I will remove this sentence.

Reviewer: OK, but then we need to make it clearer that some regions are predefined. When I first read this, I assumed that nothing is defined by default. I'll make some suggestions.

Defining a region
-----------------

Each region is defined by a name, a start address, a size, a placement strategy, and, if applicable, a device name.
A region only specifies a device name if there is a device driver associated with the region, for example, a driver for an external SPI flash.

There are three types of placement strategies, which affect how partitions are placed in regions:

start_to_end
   Place partitions sequentially from start to end.
   Partitions stored in a region with this placement strategy cannot affect their placement through the ``placement`` property.
   The unused part of the region is assigned to a partition with the same name as the region.

end_to_start
   Place partitions sequentially from end to start.
   Partitions stored in a region with this placement strategy cannot affect their placement through the ``placement`` property.
   The unused part of the region is exposed through a partition with the same name as the region.

complex
   Place partitions according to their ``placement`` configuration.
   By default, the unused part of the region is exposed through a partition named ``app``.

>    This name can be changed through a command line argument.

Reviewer: How? Can we point somewhere for more information?

Author: Hm, this feature is actually not used now, and I see no reason why we should document it. I will change it to say that the unused part is called `app`.

Regions are defined in :file:`partition_manager.cmake` as follows:

.. code-block:: cmake

  add_region(     # Define region without device name
    otp           # Name
    756           # Size
    0xff8108      # Base address
    start_to_end  # Placement strategy
    )

  add_region_with_dev(           # Define region with device name
    flash_primary                # Name
    ${flash_size}                # Size
    ${CONFIG_FLASH_BASE_ADDRESS} # Base address
    complex                      # Placement strategy
    NRF_FLASH_DRV_NAME           # Device name
    )

External flash
==============

The Partition Manager supports partitions in external flash memory through the use of :ref:`pm_regions`.
Any placeholder partition can specify that it should be stored in the external flash region.

> External flash regions always use the start_to_end placement strategy.

Reviewer: Always or by default?

Author: Always, unless you modify the build system code. Which I guess we don't document.

To use external flash, you must provide information about the device to the Partition Manager through these Kconfig options:

* :option:`CONFIG_PM_EXTERNAL_FLASH` - enable external flash

> * :option:`CONFIG_PM_EXTERNAL_FLASH_DEV_NAME` - specify the name of the flash device

Reviewer: CONFIG_PM_EXTERNAL_FLASH**_DEV**_NAME - correct?

Author: Yes!

   #endif /* CONFIG_SOC_NRF9160 */

.. _pm_regions:

Regions
=======

The Partition Manager places partitions in different *regions*.
For example, you can use regions for internal flash memory and external flash memory.

To define in which region a partition should be placed, use the ``region`` property in the configuration of the partition.

> If no region is specified, the internal flash region is used.

Reviewer: ```suggestion
If no region is specified, the predefined internal flash region is used.
```

Defining a region
-----------------

Each region is defined by a name, a start address, a size, a placement strategy, and, if applicable, a device name.
A region only specifies a device name if there is a device driver associated with the region, for example, a driver for an external SPI flash.

There are three types of placement strategies, which affect how partitions are placed in regions:

start_to_end
   Place partitions sequentially from start to end.
   Partitions stored in a region with this placement strategy cannot affect their placement through the ``placement`` property.
   The unused part of the region is assigned to a partition with the same name as the region.

end_to_start
   Place partitions sequentially from end to start.
   Partitions stored in a region with this placement strategy cannot affect their placement through the ``placement`` property.
   The unused part of the region is exposed through a partition with the same name as the region.

complex
   Place partitions according to their ``placement`` configuration.
   The unused part of the region is exposed through a partition named ``app``.

> Regions are defined in :file:`partition_manager.cmake` as follows:

Reviewer: ```suggestion
Regions are defined in :file:`partition_manager.cmake`.
For example, see the following defintions for default regions:
```

Defining a region
-----------------

Each region is defined by a name, a start address, a size, a placement strategy, and, if applicable, a device name.
A region only specifies a device name if there is a device driver associated with the region, for example, a driver for an external SPI flash.

There are three types of placement strategies, which affect how partitions are placed in regions:

start_to_end
   Place partitions sequentially from start to end.
   Partitions stored in a region with this placement strategy cannot affect their placement through the ``placement`` property.
   The unused part of the region is assigned to a partition with the same name as the region.

end_to_start
   Place partitions sequentially from end to start.
   Partitions stored in a region with this placement strategy cannot affect their placement through the ``placement`` property.
   The unused part of the region is exposed through a partition with the same name as the region.

complex
   Place partitions according to their ``placement`` configuration.
   The unused part of the region is exposed through a partition named ``app``.

Regions are defined in :file:`partition_manager.cmake`.

> For example, see the following defintions for default regions:

Reviewer: ```suggestion
For example, see the following definitions for default regions:
```

Sorry! Typo in my suggestion.

   #endif /* CONFIG_SOC_NRF9160 */

.. _pm_regions:

Regions
=======
The Partition Manager places partitions in different on *regions*.
Each region is defined by a *start address*, a *size*, a *placement strategy* and conditionally a *device name*.
Examples of regions are *internal flash* and *external flash*.
A region only specifies a *device name* if there is a device driver associated with the region, for instance a driver for an external SPI flash.
A partition sets its region using the ``region`` property in its configuration.
The internal flash region is used if no region is specified.
There are three types of *placement strategies*:

#. Start to end
   Place partitions in a start to end fashion.
   Partitions stored in a region with this placement strategy can not affect its placement through the ``placement`` keyword.

>    The unused part of the region is exposed through a partition with the same name as the region.

Reviewer 2: What does this sentence mean? "exposed" how?

Author: exposed through the defines generated by the partition manager script. 

Maybe this would be better:
```
A partition with the same name as the region is created which spans all unused area in the region.
/* or */
The unused part of the region is stored in a partition with the same name as the region.
```

   The unused part of the region is exposed through a partition with the same name as the region.

#. Complex
   Place partitions according to their ``placement`` configuration.
   The unused part of the region is exposed through a partition named ``app`` by default.
   This value can be changed through a command line argument.

Regions are defined in :file:`partition_manager.cmake` as shown below:

.. code-block:: cmake

  add_region(     # Define region without device name
    otp           # Name
    756           # Size
    0xff8108      # Base address
    start_to_end  # Placement strategy
    )

  add_region_with_dev(           # Define region with device name
    flash_primary                # Name
    ${flash_size}                # Size
    ${CONFIG_FLASH_BASE_ADDRESS} # Base address
    complex                      # Placement strategy
    NRF_FLASH_DRV_NAME           # Device name
    )

External flash
==============
The Partition Manager supports external flash through the use of :ref:`pm_regions`.
Any placeholder partition can specify that it should be stored in the external flash region.

.. note::

>   Placeholder partitions stored in the external flash region can not specify a ``placement`` configuration.

Reviewer 2: To avoid duplicating this info, say something like

> external flash has the placement strategy :ref:\`pm_placement_start_to_end\`.

The external flash information is provided to the Partition Manager through three Kconfig options.

Given the following code in the flash driver:

.. code-block:: c

>    /* SNIP */

Reviewer 2: Nit: Is `/* SNIP */` convention?

Author: No, and I guess people understand that this is not an entire C file xD removed it.
